### ISSUE TITLE
[css-typed-om] Provide a way to resolve a CSSUnitValue and CSSMath* to a pixel value

### ISSUE THREAD

**Issue by surma** [w3c/css-houdini-drafts#798]
Especially in conjunction with css-layout-api, it seems desirable to be able to resolve lengths with percentages in them. 

```js
const marginLeft = child.styleMap.get('margin-left'); // "calc(10px + 2%)"
const marginLeftResolved = CSS.resolveLength(marginLeft, {percentResolutionSize: 250});
```

... or something. A similar use case might be `column-gap` etc.

**Comment by tabatkins** [comment]
If you have the percent resolution size, you can just do this yourself, no? Just go thru and replace all the %s with an appropriate pixel value.  I suppose this is mostly a convenience method?

I'd also like to be able to resolve `em` and such, which wants to know and about the element and the property. Like, without that info, we can resolve `em` the way MQs work; with an element, we can properly resolve against 'font-size'; with an element and a property, we can correctly handle `em` in font-size vs in other properties.

We already have to `to()` method for converting a value into a specific unit; we can give it some additional powers with an options bag giving resolution details:

```
partial interface CSSNumericValue {
    CSSUnitValue to(USVString unit, CSSUnitResolvingOptions options);
};

dictionary CSSUnitResolvingOptions {
  CSSUnitValue? hundredPercent = null;
  Element? element = null;
  USVString? property = null;
};
```

And it'll succeed if it can use the provided information to resolve everything to your desired type.

**Comment by tabatkins** [comment]
(I'm using `to()` here, rather than the method shape you suggested, because we need to resolve it into some specific unit. We don't have a type that represents a nebulous "length, in some unspecified unit" type, which is the only thing we could reasonably return from such a conversion function.)

That said! I think we might want to allow something that, for example, keeps %s as they are while resolving the rest into a length as much as possible. This is useful for cases like `background-position`, where the % doesn't resolve into a length in a straightforward way. I think we can handle all the use-cases for this by just adding this options bag to `toSum()` as well; if a % doesn't trivially resolve into a length, then it can't be used in a `px / %` expression anyway, so requiring that it resolve into a sum should be fine.

In other words, calling `len.toSum("px", "percent", {element:el})` should be able to *always* succeed, and will always return a `CSSMathSum` where the first element is a `CSS.px()` and the second is a `CSS.percent()`.

**Comment by tabatkins** [comment]
Oh, hm, slight wrinkle in getting it to invoke the MQ-ish behavior. Can't just rely on not passing an element, because our *current* behavior for not passing the options bag at all (which has to be treated identically) is to *throw* when you can't convert, rather than relying on global information.

However, if I remove the default values from the dictionary listed up above, then I can test for presence/absence of the key, and use an explicit `{element: null}` as a signal to resolve based on global info like MQs do.  Leaving it out (or explicitly passing `{element:undefined}`) will leave us with the current behavior, where conversion just fails.

**Comment by surma** [comment]
I like the idea with `to()`. I was thinking of `em`s as well, actually, just forgot to put it in the OP 🤦‍♂️ 

I think your suggestion with `{element: null}` to distinguish the behaviors is fine.

Ship it? :D

**Comment by MadeByMike** [comment]
I'd love to be able to define a custom resolution. For example resolve a font-size relative to the width of a container. Or resolve a text color relative to the background. I know this stuff can be done after the initial render but it would be nice if houdini provided some way to do this within render pipeline.

**Comment by Rolf-B** [comment]
It's more complicated than that. Assume something like `transform:translate(10px,20px) translate(30%,200%) rotate(20deg) translate(-30%,-200%)` - this is from an attempt to make a card game where I show a hand of cards, then I want to drag one card away from it's position. Now I need the transformation matrix. I can to that with window.getComputedStyle(card), but there I get a string notation of the matrix and must parse it. With card.computedStyleMap().transform, I get a CSSTransformValue, but I cannot call toMatrix on it because of the % values. Converting percentages by myself means that I need to run throug all transform components, check for units that are incompatible with px, find their context, convert them according to context, et cetera. From window.getComputedStyle, I get this for free, just in string format. 
So, yes, it's convenience. A big, whopping, VERY useful convenience.

**Comment by tabatkins** [comment]
Ooh, that would be a little bit more complicated, since you want the resolution method on CSSTransformValue as well as just CSSNumericValue. That's doable, tho, and presumably in the same way.

**Comment by tabatkins** [comment]
Draft proposal: Add a .resolveUnits() method to CSSNumericValue, CSSTransformValue, and CSSTransformComponent. (Maybe CSSColorValue as well?) Defined as:

```webidl
partial interface CSSNumericValue {
    CSSUnitValue resolveUnits(CSSUnitResolvingOptions options);
};
/* etc for the other interfaces */

dictionary CSSUnitResolvingOptions {
  CSSUnitValue? hundredPercent = null;
  Element? element = null;
  USVString? property = null;
};
```

* If `hundredPercent` is specified, all percentages resolve against that.
* If `element` is specified, all non-percent relative units resolve against that element, in a generic fashion. (So 'em' uses the 'font-size' on the element, rather than the special behavior that 'em' has in 'font-size' itself.)
* If `element` *and* `property` is specified, all relative units resolve against that element *in the context of that property*. Need to define this generically in some way.
* For the transform types (and other specific types), `element` (and `property`, if needed) will resolve %s as appropriate, too, if you're not overriding it with `hundredPercent`. So you can `el.styleMap.get('transform').resolveUnits(el)` and get it fully resolved correctly, with the x and y values using the correct length to resolve %s automatically.
